app/src/lib/trampoline/trampoline-server.ts

```diff
@@ -0,0 +1,179 @@
+import { createServer, AddressInfo, Server, Socket } from 'net'
+import split2 from 'split2'
+import { enableDesktopTrampoline } from '../feature-flag'
+import { sendNonFatalException } from '../helpers/non-fatal-exception'
+import { askpassTrampolineHandler } from './trampoline-askpass-handler'
+import {
+  ITrampolineCommand,
+  TrampolineCommandHandler,
+  TrampolineCommandIdentifier,
+} from './trampoline-command'
+import { TrampolineCommandParser } from './trampoline-command-parser'
+import { isValidTrampolineToken } from './trampoline-tokens'
+
+/**
+ * This class represents the "trampoline server". The trampoline is something
+ * we'll hand to git in order to communicate with Desktop without noticing. A
+ * notable example of this would be GIT_ASKPASS.
+ *
+ * This server is designed so that it will start lazily when the app performs a
+ * remote git operation. At that point, the app will try to retrieve the
+ * server's port, which will run the server first if needed.
+ *
+ * The idea behind this is to simplify the retry approach in case of error:
+ * instead of reacting to errors with an immediate retry, the server will remain
+ * closed until the next time the app needs it (i.e. in the next git remote
+ * operation).
+ */
+export class TrampolineServer {
+  private readonly server: Server
+  private listeningPromise: Promise<void> | null = null
+
+  private readonly commandHandlers = new Map<
+    TrampolineCommandIdentifier,
+    TrampolineCommandHandler
+  >()
+
+  public constructor() {
+    this.server = createServer(socket => this.onNewConnection(socket))
+
+    this.registerCommandHandler(
+      TrampolineCommandIdentifier.AskPass,
+      askpassTrampolineHandler
+    )
+  }
+
+  private async listen(): Promise<void> {
+    if (!enableDesktopTrampoline()) {
+      this.listeningPromise = Promise.resolve()
+      return this.listeningPromise
+    }
+
+    this.listeningPromise = new Promise((resolve, reject) => {
+      // Observe errors while trying to start the server
+      this.server.on('error', error => {
+        reject(error)
+        this.close()
+      })
+
+      this.server.listen(0, '127.0.0.1', async () => {
```

Comment:
Will the TCP server be accessible to other computers on the network?

Reply:
We definitely don't want it to be, and that `'127.0.0.1'` should restrict it to connections in the same machine (or that's my hope, at least 😂 ).

Reply:
Let me test and see if it's really keeping the connections to local or if I can use vpn from my company to connect